File: renderers/ios/VarnFilePicker.swift
import PhotosUI
import UIKit
import UniformTypeIdentifiers

/// The button that opens the chooser the system ships and reports what came back.
///
/// Two different controllers stand behind it. A picture comes from the photo library through
/// `PHPickerViewController`, which needs no permission at all because the reader chooses inside a view
/// the system owns, and anything else comes from the files app through `UIDocumentPickerViewController`.
/// A file chosen there is reached inside a security-scoped resource, which has to be given up again or
/// the process leaks the access for the life of the app.
final class VarnFilePicker: UIButton {
    private var kind = "file"
    private var accept: [String] = []
    private var multiple = false
    private(set) var limit = 8 * 1024 * 1024

    var onPick: (([String: Any]) -> Void)?

    override init(frame: CGRect) {
        super.init(frame: frame)
        addTarget(self, action: #selector(open), for: .touchUpInside)
    }

    required init?(coder: NSCoder) { fatalError("not used") }

    func setKind(_ value: String?) {
        kind = value ?? "file"
    }

    func setAccept(_ value: [String]) {
        accept = value
    }

    func setMultiple(_ value: Bool) {
        multiple = value
    }

    func setMaxBytes(_ value: Int) {
        limit = value
    }

    /// Answers the types the chooser offers, which is everything when the tree named none.
    ///
    /// An entry with a slash in it is a mime type and anything else is an extension, since those are the
    /// two ways a caller writes one and neither is worth a prop of its own.
    private var types: [UTType] {
        let named = accept.compactMap { entry -> UTType? in
            entry.contains("/") ? UTType(mimeType: entry) : UTType(filenameExtension: entry)
        }

        if !named.isEmpty {
            return named
        }

        return kind == "image" ? [.image] : [.item]
    }

    @objc private func open() {
        guard let presenter = window?.rootViewController else {
            return
        }

        if kind == "image" {
            var settings = PHPickerConfiguration()

            settings.filter = .images
            settings.selectionLimit = multiple ? 0 : 1

            let picker = PHPickerViewController(configuration: settings)
            picker.delegate = self
            presenter.present(picker, animated: true)
            return
        }

        let picker = UIDocumentPickerViewController(forOpeningContentTypes: types, asCopy: true)

        picker.allowsMultipleSelection = multiple
        picker.delegate = self
        presenter.present(picker, animated: true)
    }

    /// Answers one chosen file as the four fields every renderer reports, reading it only within the cap.
    func entry(name: String, type: String, data: @autoclosure () -> Data?) -> [String: Any] {
        guard let bytes = data() else {
            return ["name": name, "size": 0, "type": type, "bytes": NSNull()]
        }

        guard bytes.count <= limit else {
            return ["name": name, "size": bytes.count, "type": type, "bytes": NSNull()]
        }

        return [
            "name": name,
            "size": bytes.count,
            "type": type,
            "bytes": bytes.base64EncodedString(),
        ]
    }

    /// Reports one chosen file, as soon as it has been read.
    ///
    /// Gathering the whole set first put the bytes of every photograph into one message, which is
    /// decoded and written in a single turn of the loop: nothing moved until the last one landed.
    private func report(_ file: [String: Any]) {
        onPick?(file)
    }
}

extension VarnFilePicker: UIDocumentPickerDelegate {
    func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        for url in urls {
            let reached = url.startAccessingSecurityScopedResource()

            let type = UTType(filenameExtension: url.pathExtension)?.preferredMIMEType
                ?? "application/octet-stream"

            report(entry(name: url.lastPathComponent, type: type, data: try? Data(contentsOf: url)))

            if reached {
                url.stopAccessingSecurityScopedResource()
            }
        }
    }
}

extension VarnFilePicker: PHPickerViewControllerDelegate {
    func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
        picker.dismiss(animated: true)

        guard !results.isEmpty else {
            return
        }

        for (index, result) in results.enumerated() {
            let provider = result.itemProvider

            // The library names a picture only sometimes, and several chosen at once would all arrive
            // called the same thing — which is one file written over and over for anything that keeps
            // them by name. The type is the one the item actually carries rather than a guess.
            let carried = provider.registeredTypeIdentifiers.compactMap(UTType.init).first { $0.conforms(to: .image) }
            let type = carried ?? .jpeg
            let extension_ = type.preferredFilenameExtension ?? "jpg"
            let name = provider.suggestedName.map { "\($0).\(extension_)" } ?? "picture \(index + 1).\(extension_)"

            provider.loadDataRepresentation(forTypeIdentifier: type.identifier) { [weak self] data, _ in
                guard let self else {
                    return
                }

                let file = self.entry(
                    name: name,
                    type: type.preferredMIMEType ?? "image/jpeg",
                    data: data
                )

                DispatchQueue.main.async { self.report(file) }
            }
        }
    }
}

File: renderers/android/src/main/kotlin/dev/varn/gui/VarnFilePicker.kt
package dev.varn.gui

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.OpenableColumns
import android.util.Base64
import android.widget.Button
import org.json.JSONObject

/**
 * The button that opens the chooser the system ships and reports what came back.
 *
 * A chooser is an activity, and an activity's result belongs to the activity that started it, so the
 * button does not start one itself: it asks the host, the host asks the application, and the
 * application hands the result back the same way it hands over the back press. Nothing here holds an
 * activity, which is what would leak one on a rotation.
 */
class VarnFilePicker(context: Context) : Button(context) {
    private var kind = "file"
    private var accept: List<String> = emptyList()
    private var multiple = false
    private var limit = 8 * 1024 * 1024

    /** Asked with the intent to start, and answered later through [chosen]. */
    var onOpen: ((VarnFilePicker, Intent) -> Unit)? = null

    var onPick: ((JSONObject) -> Unit)? = null

    init {
        setOnClickListener { open() }
    }

    fun setKind(value: String?) {
        kind = value ?: "file"
    }

    fun setAccept(value: List<String>) {
        accept = value
    }

    fun setMultiple(value: Boolean) {
        multiple = value
    }

    fun setMaxBytes(value: Int) {
        limit = value
    }

    /**
     * Answers the types the chooser offers, which is everything when the tree named none.
     *
     * An entry with a slash in it is a mime type and anything else is an extension, since those are the
     * two ways a caller writes one and neither is worth a prop of its own.
     */
    private fun types(): Array<String> {
        val named = accept.map { entry ->
            if (entry.contains("/")) entry else "application/${entry.trimStart('.')}"
        }

        if (named.isNotEmpty()) {
            return named.toTypedArray()
        }

        return if (kind == "image") arrayOf("image/*") else arrayOf("*/*")
    }

    private fun open() {
        val offered = types()
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)

        intent.addCategory(Intent.CATEGORY_OPENABLE)
        intent.type = offered.first()
        intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, multiple)

        if (offered.size > 1) {
            intent.putExtra(Intent.EXTRA_MIME_TYPES, offered)
        }

        onOpen?.invoke(this, intent)
    }

    /**
     * Reads what was chosen and reports each one as it is ready.
     *
     * Gathering the whole set first put the bytes of every photograph into one message, which is decoded
     * and written in a single turn of the loop: nothing moved until the last one landed.
     */
    fun chosen(uris: List<Uri>) {
        for (uri in uris) {
            onPick?.invoke(entry(uri))
        }
    }

    private fun entry(uri: Uri): JSONObject {
        val resolver = context.contentResolver
        var name = uri.lastPathSegment ?: "file"
        var size = 0L

        resolver.query(uri, null, null, null, null)?.use { row ->
            val named = row.getColumnIndex(OpenableColumns.DISPLAY_NAME)
            val measured = row.getColumnIndex(OpenableColumns.SIZE)

            if (row.moveToFirst()) {
                if (named >= 0) {
                    name = row.getString(named)
                }

                if (measured >= 0 && !row.isNull(measured)) {
                    size = row.getLong(measured)
                }
            }
        }

        val entry = JSONObject()
            .put("name", name)
            .put("size", size)
            .put("type", resolver.getType(uri) ?: "application/octet-stream")

        if (size > limit) {
            return entry.put("bytes", JSONObject.NULL)
        }

        val bytes = runCatching { resolver.openInputStream(uri)?.use { it.readBytes() } }.getOrNull()

        if (bytes == null || bytes.size > limit) {
            return entry.put("size", bytes?.size ?: size).put("bytes", JSONObject.NULL)
        }

        return entry.put("size", bytes.size).put("bytes", Base64.encodeToString(bytes, Base64.NO_WRAP))
    }
}
